### DISCUSSION TITLE
How to handle BodyStream properly, when you don't receive it but return HTTP error e.g. 409

### DISCUSSION THREAD

**Discussion by atimin** [tokio-rs/axum#2016] [Q&A] [answered]
Hey everyone,

I'm working on a database with an HTTP frontend based on axum. I met some issue when I handle POST request with BodyStream. My handler checks many parameters and if the incoming data  is valid it receives the body. If something is wrong it returns an error:

```rust
    pub async fn write_record(
        State(components): State<Arc<RwLock<HttpServerComponents>>>,
        headers: HeaderMap,
        Path(path): Path<HashMap<String, String>>,
        Query(params): Query<HashMap<String, String>>,
        mut stream: BodyStream,
    ) -> Result<(), HttpError> {
        let bucket = path.get("bucket_name").unwrap();
        check_permissions(
            Arc::clone(&components),
            headers.clone(),
            WriteAccessPolicy {
                bucket: bucket.clone(),
            },
        )?;

        if !params.contains_key("ts") {
            return Err(HttpError::unprocessable_entity(
                "'ts' parameter is required",
            ));
        }

      // many other checks
        while let Some(chunk) = stream.next().await {
            let mut writer = writer.write().unwrap();
            let chunk = match chunk {
                Ok(chunk) => chunk,
                Err(e) => {
                    writer.write(Chunk::Error)?;
                    error!("Error while receiving data: {}", e);
                    return Err(HttpError::from(e));
                }
            };
            writer.write(Chunk::Data(chunk))?;
        }

        writer.write().unwrap().write(Chunk::Last(Bytes::new()))?;
        Ok(())
}
```
It works perfectly if there is no error happening, but if the server returns an HTTP error, on the client side I have sometimes (!!!) an communication error (broken pipe). Looks like the client is still trying to stream the data but the server interrupts it.  I feel that I have to handle the case with BodyStream in a special manner but I don't see how.

**Comment by davidpdrsn** [comment]
Returning a server error (such as 500) doesn't interrupt the response stream. Sounds like an issue with your client?

**Reply by atimin** [reply]
Thanks for the response!
I've tried two different client libraries... I've rewritten the database from C++ to Rust. The C++ implementation didn't have such a problem. But if you say it should not interrupt the stream , then I have a bug somewhere .

**Reply by davidpdrsn** [reply]
Interrupt which stream? Client to server or server to client? You're talking about sending server errors and getting client disconnects. I don't follow.

**Reply by atimin** [reply]
It is a post request.  A client streams data to a server. The server checks the correctness of the request and returns an error if it is invalid.  If the request is OK, the server receives the stream.

**Reply by atimin** [reply]
So it's a request stream..

**Comment by atimin** [comment, marked as answer]
The issue was fixed after I  added  the "Expect:100-continue" header to a request on the client side.

**Reply by atimin** [reply]
However, it doesn't work out of the box for many HTTP clients. You have to implement the "Expect" logic according to [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-5.1.1)
